src/serializer/ResidualHeapVisitor.js

```diff
@@ -780,10 +780,12 @@ export class ResidualHeapVisitor {
     } else if (val instanceof FunctionValue) {
       // Function declarations should get hoisted in common scope so that instances only get allocated once
       let parentScope = this.scope;
-      this._withScope(this.commonScope, () => {
-        invariant(val instanceof FunctionValue);
-        if (this.preProcessValue(val)) this.visitValueFunction(val, parentScope);
-      });
+      // Every function references itself through arguments, prevent the recursive double-visit
```

Comment:
My first thought about this was that it is still broken because it does not deal with other forms of recursion. Then I remembered that the abstract interpreter cannot deal with it either (and will give a diagnostic), so no problem with that.